mcp/server.go

```diff
@@ -138,59 +140,100 @@ func (s *Server) RemovePrompts(names ...string) {
 		func() bool { return s.prompts.remove(names...) })
 }
 
-// AddTool adds a [Tool] to the server, or replaces one with the same name.
+// AddRawTool adds a [Tool] to the server, or replaces one with the same name.
 // The Tool argument must not be modified after this call.
 //
 // The tool's input schema must be non-nil. For a tool that takes no input,
 // or one where any input is valid, set [Tool.InputSchema] to the empty schema,
 // &jsonschema.Schema{}.
-func (s *Server) AddTool(t *Tool, h ToolHandler) {
-	if t.InputSchema == nil {
-		// This prevents the tool author from forgetting to write a schema where
-		// one should be provided. If we papered over this by supplying the empty
-		// schema, then every input would be validated and the problem wouldn't be
-		// discovered until runtime, when the LLM sent bad data.
-		panic(fmt.Sprintf("adding tool %q: nil input schema", t.Name))
-	}
-	if err := addToolErr(s, t, h); err != nil {
-		panic(err)
-	}
+//
+// When the handler is invoked as part of a CallTool request, req.Params.Arguments
+// will be a json.RawMessage. Unmarshaling the arguments and validating them against the
+// input schema are the handler author's responsibility.
+func (s *Server) AddRawTool(t *Tool, h RawToolHandler) {
+	st := &serverTool{tool: t, handler: h}
+	// Assume there was a change, since add replaces existing tools.
+	// (It's possible a tool was replaced with an identical one, but not worth checking.)
+	// TODO: Batch these changes by size and time? The typescript SDK doesn't.
+	// TODO: Surface notify error here? best not, in case we need to batch.
+	s.changeAndNotify(notificationToolListChanged, &ToolListChangedParams{},
+		func() bool { s.tools.add(st); return true })
 }
 
-// AddTool adds a [Tool] to the server, or replaces one with the same name.
 // If the tool's input schema is nil, it is set to the schema inferred from the In
 // type parameter, using [jsonschema.For].
 // If the tool's output schema is nil and the Out type parameter is not the empty
 // interface, then the output schema is set to the schema inferred from Out.
-// The Tool argument must not be modified after this call.
-func AddTool[In, Out any](s *Server, t *Tool, h ToolHandlerFor[In, Out]) {
-	if err := addToolErr(s, t, h); err != nil {
-		panic(err)
+func RawToolHandlerFor[In, Out any](t *Tool, h ToolHandlerFor[In, Out]) RawToolHandler {
```

Comment:
When used like this, there's no way to chain the call, which seemed critical.

Isn't it better not to mutate the input?